File: packages/pvm-assemblyscript/assembly/host-functions/accumulate/bless.ts
import { RESULT_CODE_PANIC } from '../../config'
import { AlwaysAccerEntry } from '../../codec'
import {
  ACCUMULATE_ERROR_WHAT,
  ACCUMULATE_ERROR_WHO,
  AccumulateHostFunctionContext,
  BaseAccumulateHostFunction,
  HostFunctionResult,
} from './base'

/**
 * BLESS accumulation host function (Ω_B)
 *
 * Empowers service with manager, assigners, delegator, registrar, and always accessors
 *
 * Gray Paper Specification:
 * - Function ID: 14 (bless)
 * - Gas Cost: 10
 * - Parameters: registers[7-12] = m, a, v, r, o, n
 *   - m: manager service ID
 *   - a: assigners array offset in memory
 *   - v: delegator service ID
 *   - r: registrar service ID
 *   - o: always accessors array offset in memory
 *   - n: number of always accessors
 * - Returns: registers[7] = OK or error code
 *
 * Gray Paper Logic:
 * 1. Read assigners array from memory (341 cores * 4 bytes each)
 * 2. Read always accessors array from memory (n entries * 12 bytes each)
 * 3. Validate service IDs are valid
 * 4. Update state with new privileges
 */
export class BlessHostFunction extends BaseAccumulateHostFunction {
  functionId: u64 = u64(14) // BLESS function ID
  name: string = 'bless'
  gasCost: u64 = u64(10)

  // Gray Paper constants
  C_CORE_COUNT: u64 = u64(341) // Ccorecount (default)

  execute(context: AccumulateHostFunctionContext): HostFunctionResult {
    const registers = context.registers
    const ram = context.ram
    const implications = context.implications
    // Extract parameters from registers
    const managerServiceId = u64(registers[7])
    const assignersOffset = u64(registers[8])
    const delegatorServiceId = u64(registers[9])
    const registrarServiceId = u64(registers[10])
    const alwaysAccessorsOffset = u64(registers[11])
    const numberOfAlwaysAccessors = u64(registers[12])

    // Read assigners array from memory (341 cores * 4 bytes each)
    const assignersLength = this.C_CORE_COUNT * u64(4)
    const readResult_assigners = ram.readOctets(
      u32(assignersOffset),
      u32(assignersLength),
    )
    if (readResult_assigners.faultAddress !== 0) {
      this.setAccumulateError(registers, ACCUMULATE_ERROR_WHAT)
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }
    if (readResult_assigners.data === null) {
      this.setAccumulateError(registers, ACCUMULATE_ERROR_WHAT)
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }
    const assignersData = readResult_assigners.data!

    // Read always accessors array from memory (n entries * 12 bytes each)
    const accessorsLength = numberOfAlwaysAccessors * u64(12)
    const readResult_accessors = ram.readOctets(
      u32(alwaysAccessorsOffset),
      u32(accessorsLength),
    )
    if (readResult_accessors.faultAddress !== 0) {
      this.setAccumulateError(registers, ACCUMULATE_ERROR_WHAT)
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }
    if (readResult_accessors.data === null) {
      this.setAccumulateError(registers, ACCUMULATE_ERROR_WHAT)
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }
    const accessorsData = readResult_accessors.data!

    // Parse assigners array (4 bytes per core ID, little-endian)
    const assigners: u32[] = []
    for (let i: i32 = 0; i < i32(this.C_CORE_COUNT); i++) {
      const offset = i * 4
      // Read 4 bytes as little-endian u32
      let coreId: u32 = u32(0)
      coreId |= u32(assignersData[offset])
      coreId |= u32(assignersData[offset + 1]) << 8
      coreId |= u32(assignersData[offset + 2]) << 16
      coreId |= u32(assignersData[offset + 3]) << 24
      assigners.push(coreId)
    }

    // Parse always accessors array (12 bytes per accessor: 4 bytes service ID + 8 bytes gas, little-endian)
    const alwaysAccessors: Array<AlwaysAccerEntry> = []
    for (let i: i32 = 0; i < i32(numberOfAlwaysAccessors); i++) {
      const offset = i * 12
      // Read service ID (4 bytes, little-endian)
      let serviceId: u32 = u32(0)
      serviceId |= u32(accessorsData[offset])
      serviceId |= u32(accessorsData[offset + 1]) << 8
      serviceId |= u32(accessorsData[offset + 2]) << 16
      serviceId |= u32(accessorsData[offset + 3]) << 24
      // Read gas (8 bytes, little-endian)
      let gas: u64 = u64(0)
      gas |= u64(accessorsData[offset + 4])
      gas |= u64(accessorsData[offset + 5]) << 8
      gas |= u64(accessorsData[offset + 6]) << 16
      gas |= u64(accessorsData[offset + 7]) << 24
      gas |= u64(accessorsData[offset + 8]) << 32
      gas |= u64(accessorsData[offset + 9]) << 40
      gas |= u64(accessorsData[offset + 10]) << 48
      gas |= u64(accessorsData[offset + 11]) << 56
      alwaysAccessors.push(new AlwaysAccerEntry(serviceId, gas))
    }

    // Validate service IDs
    // Gray Paper line 706: (m, v, r) not in serviceid^3 → return WHO
    // serviceid ≡ Nbits{32} (Gray Paper definitions.tex line 15, accounts.tex line 7)
    // So each service ID must be: 0 ≤ id < 2^32
    const MAX_SERVICE_ID: u64 = u64(4294967296) // 2^32
    const isValidServiceId = (id: u64): bool => {
      return id < MAX_SERVICE_ID
    }

    if (
      !isValidServiceId(managerServiceId) ||
      !isValidServiceId(delegatorServiceId) ||
      !isValidServiceId(registrarServiceId)
    ) {
      this.setAccumulateError(registers, ACCUMULATE_ERROR_WHO)
      return new HostFunctionResult(255) // continue execution
    }

    // Update state with new privileges
    // Gray Paper: imX.state = {manager: m, assigners: a, delegator: v, registrar: r, alwaysaccers: z}
    const imX = implications.regular
    imX.state.manager = u32(managerServiceId)
    imX.state.assigners = assigners
    imX.state.delegator = u32(delegatorServiceId)
    imX.state.registrar = u32(registrarServiceId)
    imX.state.alwaysaccers = alwaysAccessors

    // Set success result
    this.setAccumulateSuccess(registers)
    return new HostFunctionResult(255) // continue execution
  }
}

File: packages/pvm-rust/src/host_functions/accumulate/bless.rs
//! BLESS accumulation host function (Ω_B). Gray Paper: function ID 14.
//! r7..r12 = m, a, v, r, o, n (manager, assigners offset, delegator, registrar, always accessors offset, n). 1:1 with AS bless.ts.

use crate::codec::AlwaysAccerEntry;
use crate::config::FUNC_BLESS;
use crate::host_functions::accumulate::base::{self, codes, DEFAULT_NUM_CORES, MAX_SERVICE_ID};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

const BYTES_PER_ASSIGNER: u32 = 4;
const BYTES_PER_ALWAYS_ACCER: u32 = 12;

pub struct BlessHostFunction;

impl BlessHostFunction {
    fn parse_assigners(data: &[u8], num_cores: u32) -> Vec<u32> {
        let mut assigners = Vec::with_capacity(num_cores as usize);
        for i in 0..(num_cores as usize) {
            let offset = i * BYTES_PER_ASSIGNER as usize;
            let core_id = u32::from_le_bytes([
                data[offset],
                data[offset + 1],
                data[offset + 2],
                data[offset + 3],
            ]);
            assigners.push(core_id);
        }
        assigners
    }

    fn parse_always_accessors(data: &[u8], n: u64) -> Vec<AlwaysAccerEntry> {
        let mut entries = Vec::with_capacity(n as usize);
        for i in 0..(n as usize) {
            let offset = i * BYTES_PER_ALWAYS_ACCER as usize;
            let service_id = u32::from_le_bytes([
                data[offset],
                data[offset + 1],
                data[offset + 2],
                data[offset + 3],
            ]);
            let gas = u64::from_le_bytes([
                data[offset + 4],
                data[offset + 5],
                data[offset + 6],
                data[offset + 7],
                data[offset + 8],
                data[offset + 9],
                data[offset + 10],
                data[offset + 11],
            ]);
            entries.push(AlwaysAccerEntry { service_id, gas });
        }
        entries
    }
}

impl HostFunction for BlessHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_BLESS
    }
    fn name(&self) -> &'static str {
        "bless"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let manager_service_id = context.registers[7];
        let assigners_offset = context.registers[8];
        let delegator_service_id = context.registers[9];
        let registrar_service_id = context.registers[10];
        let always_accessors_offset = context.registers[11];
        let number_of_always_accessors = context.registers[12];

        let num_cores = context
            .num_cores
            .unwrap_or(DEFAULT_NUM_CORES);
        let assigners_length = (u64::from(num_cores) * u64::from(BYTES_PER_ASSIGNER)) as u32;
        let read_assigners = context.ram.read_octets(assigners_offset as u32, assigners_length);
        if read_assigners.fault_address != 0 || read_assigners.data.is_none() {
            crate::host_log!(
                "[hostfn] bless PANIC: assigners read fault (offset={}, len={}, fault_address={})",
                assigners_offset, assigners_length, read_assigners.fault_address
            );
            return HostFunctionResult::panic();
        }
        let assigners_data = read_assigners.data.unwrap();
        if assigners_data.len() != assigners_length as usize {
            crate::host_log!(
                "[hostfn] bless PANIC: assigners length mismatch (got {}, expected {})",
                assigners_data.len(),
                assigners_length as usize
            );
            return HostFunctionResult::panic();
        }

        let accessors_length = (number_of_always_accessors * u64::from(BYTES_PER_ALWAYS_ACCER)) as u32;
        let read_accessors = context.ram.read_octets(always_accessors_offset as u32, accessors_length);
        if read_accessors.fault_address != 0 || read_accessors.data.is_none() {
            crate::host_log!(
                "[hostfn] bless PANIC: accessors read fault (offset={}, len={}, fault_address={})",
                always_accessors_offset, accessors_length, read_accessors.fault_address
            );
            return HostFunctionResult::panic();
        }
        let accessors_data = read_accessors.data.unwrap();
        if accessors_data.len() != accessors_length as usize {
            crate::host_log!(
                "[hostfn] bless PANIC: accessors length mismatch (got {}, expected {})",
                accessors_data.len(),
                accessors_length as usize
            );
            return HostFunctionResult::panic();
        }

        if manager_service_id >= MAX_SERVICE_ID
            || delegator_service_id >= MAX_SERVICE_ID
            || registrar_service_id >= MAX_SERVICE_ID
        {
            base::set_accumulate_error(context.registers, codes::WHO);
            return HostFunctionResult::continue_execution();
        }

        let assigners = Self::parse_assigners(&assigners_data, num_cores);
        let alwaysaccers = Self::parse_always_accessors(&accessors_data, number_of_always_accessors);

        // AS always has implications.regular; single imX.state (accumulation_state).
        let state = context
            .accumulation_state
            .as_deref_mut()
            .expect("bless: accumulation context must have accumulation_state");

        state.manager = manager_service_id as u32;
        state.delegator = delegator_service_id as u32;
        state.registrar = registrar_service_id as u32;
        state.assigners = assigners;
        state.alwaysaccers = alwaysaccers;

        base::set_accumulate_success(context.registers, codes::OK);
        HostFunctionResult::continue_execution()
    }
}
